### DISCUSSION TITLE
Compile libcurl with WebSockets statically

### DISCUSSION THREAD

**Discussion by TracerDS** [curl/curl#13849] [Q&A] [answered]
Hello, in [previous discussion](https://github.com/curl/curl/discussions/13844) I asked about WebSocket support.
From that I decided to compile libcurl from source statically (as I've been using static libcurl for my project. I need to replace it with the libcurl with WSS support) but I have some issues while doing so.

From [the windows compilation mini-tutorial](https://github.com/curl/curl/discussions/11075) I compiled libcurl dynamically successfully, but how to do it statically?

```bat
cd .\curl-master
cmake . -DCURL_USE_SCHANNEL=ON -DENABLE_WEBSOCKETS=ON -DCMAKE_INSTALL_PREFIX:PATH=.\cmake-install
cmake --build . --target INSTALL --config Release
```
I understand in order to build a debug version I need to change `--config Release` to `--config Debug`, but what about shared vs dynamic lib?

Is it possible to compile it statically via cmake?

**Comment by TracerDS** [comment]
After some digging I found a flag responsible for building a static lib in the CMakeList but the output lib is invalid, doesnt want to link (linking errors) and cannot be compiled with msvcrt runtime

**Comment by talregev** [comment]
I tried with vcpkg:
https://github.com/microsoft/vcpkg

```
git clone https://github.com/microsoft/vcpkg
cd vcpkg
bootstrap-vcpkg.bat
vcpkg install curl[tool,websockets]:x64-windows-static
cd installed\x64-windows-static\tools\curl
curl --version
```
Enjoy :)

**Reply by TracerDS** [reply]
> I tried with vcpkg: https://github.com/microsoft/vcpkg
> 
> ```
> git clone https://github.com/microsoft/vcpkg
> cd vcpkg
> bootstrap-vcpkg.bat
> vcpkg install curl[tool,websockets]:x64-windows-static
> cd installed\x64-windows-static\tools\curl
> curl --version
> ```
> 
> Enjoy :)

Hmm, interesting. I've successfully compiled libcurl via nmake and winbuilds but didnt know vcpkg supported this functionality too!
Does cmake also support it?

**Reply by talregev** [reply]
vcpkg compile curl with cmake. and you can add vcpkg to your cmake project, to compile your dependencies.

**Comment by vszakats** [comment, marked as answer]
Yes, CMake supports it with these two options:
```shell
-DBUILD_STATIC_LIBS=ON
-DENABLE_WEBSOCKETS=ON
```

If you want to test pre-built binaries with these options, you can try the official curl-for-win downloads at <https://curl.se/windows>. They are built with CMake and also include static libcurl.

**Reply by TracerDS** [reply]
> Yes, CMake supports it with these two options:
> 
> ```shell
> -DBUILD_STATIC_LIBS=ON
> -DENABLE_WEBSOCKETS=ON
> ```


That will just define preprocessor options, no? If so, that wont work since libcurl builds to dynamic lib by default

> If you want to test pre-built binaries with these options, you can try the official curl-for-win downloads at https://curl.se/windows. They are built with CMake and also include static libcurl.

But they dont have WebSockets included. And they dont have debug info.

**Reply by vszakats** [reply]
No, these two are CMake options, that you need to pass to CMake (as in your initial post). The top one enables static libs.

And yes, curl-for-win builds do have WebSockets enabled.

They don't have debug info though.

**Reply by TracerDS** [reply]
> And yes, curl-for-win builds do have WebSockets enabled.

Huh, didnt know about that.
Thanks! Thats exactly what I was looking for (cmake options)
